jdaviz/configs/cubeviz/plugins/parsers.py

```diff
@@ -210,12 +256,14 @@ def _parse_esa_s3d(app, hdulist, data_label, ext='DATA', viewer_name='flux-viewe
     flux = np.moveaxis(flux, 0, -1)
     flux = np.swapaxes(flux, 0, 1)
 
-    metadata = standardize_metadata(hdulist[ext].header)
+    metadata = standardize_metadata(hdu.header)
     metadata.update(wcs_dict)  # To be internally consistent
-    if hdulist[ext].name != 'PRIMARY' and 'PRIMARY' in hdulist:
+    if hdu.name != 'PRIMARY' and 'PRIMARY' in hdulist:
         metadata[PRIHDR_KEY] = standardize_metadata(hdulist['PRIMARY'].header)
 
-    data = Spectrum1D(flux, wcs=wcs, meta=metadata)
+    data = _return_spectrum_with_correct_units(flux, wcs, metadata, data_type, hdulist=hdulist)
+    if data_type == 'flux':  # Forced wave unit conversion made it lose stuff, so re-add
+        app.data_collection[-1].get_component("flux").units = flux.unit
```

Comment:
does this logic (the `[-1]`) work as expected with the `app.add_data` call _after_?  If this is intentional, an inline comment explaining the logic might be justified (intentionally changing the units of the previous data collection entry?).

Reply:
You are correct, this should be after the add to data call.

Reply:
p.s. OK now that I look at it closer, yes, you should move it. This is the ESA stuff that is not tested.